pkg/upload/js.go

```diff
@@ -16,6 +24,158 @@ type JsOptions struct {
+		return appVersion, nil
+	}
+	return "", fmt.Errorf("unable to locate package.json to resolve version in %s", projectRoot)
+}
+
+// Attempt to find the source maps by walking the build directory if it is not passed in to the command line
+func ResolveSourceMapPaths(jsOptions JsOptions, outputPath string, projectRoot string, logger log.Logger) ([]string, error) {
+	if jsOptions.SourceMap != "" {
+		sourceMap := jsOptions.SourceMap
+		if !path.IsAbs(sourceMap) {
+			sourceMap = filepath.Join(projectRoot, sourceMap)
+		}
+		if utils.FileExists(sourceMap) {
+			return []string{sourceMap}, nil
+		} else {
+			return []string{}, fmt.Errorf("unable to find specified source map file: %s", sourceMap)
+		}
+	}
+
+	var sourceMapPaths []string
+	err := filepath.WalkDir(outputPath, func(fullPath string, dirEntry fs.DirEntry, err error) error {
+		if !dirEntry.IsDir() && strings.HasSuffix(dirEntry.Name(), ".map") && !strings.HasSuffix(dirEntry.Name(), ".css.map") {
+			sourceMapPaths = append(sourceMapPaths, fullPath)
+		}
+		return nil
+	})
+	if err != nil {
+		return []string{}, err
+	}
+	if len(sourceMapPaths) == 0 {
+		logger.Debug(fmt.Sprintf("No main source map found in: %s", outputPath))
+	} else {
+		logger.Debug(fmt.Sprintf("Found source maps: %s", strings.Join(sourceMapPaths, ", ")))
+	}
+	return sourceMapPaths, nil
+}
+
+// Attempt to find the bundle path by changing the extension of the source map if the bundle path is not passed in to the command line
+func ResolveBundlePath(jsOptions JsOptions, sourceMapPath string) (string, error) {
+	if jsOptions.Bundle != "" {
+		if utils.FileExists(jsOptions.Bundle) {
+			return jsOptions.Bundle, nil
+		} else {
+			return "", fmt.Errorf("unable to find specified bundle: %s", jsOptions.Bundle)
+		}
+	}
+
+	withoutSuffix, found := strings.CutSuffix(sourceMapPath, ".map")
+	if !found {
+		return "", nil
+	}
+
+	if utils.FileExists(withoutSuffix) {
+		return withoutSuffix, nil
+	}
+	return "", nil
+}
+
+// Upload a single source map
+func UploadSingleSourceMap(
+	jsOptions JsOptions,
+	sourceMapPath string,
+	apiKey string,
+	appVersion string,
+	projectRoot string,
+	endpoint string,
+	timeout int,
+	retries int,
+	overwrite bool,
+	dryRun bool,
+	logger log.Logger,
+) {
+
+	bundlePath, err := ResolveBundlePath(jsOptions, sourceMapPath)
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	uploadOptions, err := utils.BuildJsUploadOptions(apiKey, appVersion, jsOptions.BundleUrl, projectRoot, overwrite)
+
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	fileFieldData := make(map[string]string)
+	fileFieldData["sourceMap"] = sourceMapPath
+	if bundlePath != "" {
+		fileFieldData["minifiedFile"] = bundlePath
+	}
+
+	err = server.ProcessFileRequest(endpoint+"/sourcemap", uploadOptions, fileFieldData, timeout, retries, sourceMapPath, dryRun, logger)
+
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+}
+
 func ProcessJs(
 	apiKey string,
 	jsOptions JsOptions,
```

Comment:
Passing a whole command struct isn't something we've done elsewhere – although we've talked about doing it. I'd be happier following our existing pattern of passing the individual options down for now, and we can discuss simplifying the methods by passing a single struct as a future enhancement.

Reply:
Could you elaborate on the advantage of passing each field in the struct rather than just the struct itself?

Reply:
Consistency with the existing commands in the codebase. I'd rather refactor everything at once in a separate PR than change the style for just this one command now.